### ISSUE TITLE
Feature importance should consider pruned trials

### ISSUE THREAD

**Issue by mreitschuster** [optuna/optuna#3709]
### Motivation

see also [this discussion/help request](https://github.com/optuna/optuna/discussions/3703) and [PR that updates the docs to describe current behavior](https://github.com/optuna/optuna/pull/3707).

With f1 and f2 as features/tuned hyperparameters
```
f1 == A & f2==X -> 7x PRUNED WITH VALUE 0.1
f1 == A & f2==Y -> 8x PRUNED WITH VALUE 0.1
f1 == B  & f2==X   -> 7x COMPLETE WITH VALUE 1   
f1 == B  & f2==Y   -> 8x COMPLETE WITH VALUE 1.1
```

My expectation from a feature importance analysis is that it shows that f1 is more important than f2 - if I want to maximize the value my best bet is to set f1 (to B), rather than set f2.

### Description

It is clear that assumptions for the value of a pruned trial needs to be made. There are some simple, potential choices like: last intermediate value, 0, minimum value of all observed, some interpolation /extrapolation. It might even be framed as a missing value problem (i.e. not-missing-at-random).

My feature request would be to allow the user when calling functions from optuna.importance and optuna.visualization to provide a parameter **impute_pruned** which can have values:

- **None**  behavior as today, not taking pruned into account

- **any float** impute this value 

- **function(trial) -> float**    user can hand over a function that does imputing based on the trial object, e.g. can use the last intermediate value or some calculation on top of it

### Alternatives (optional)

add function similar in intent as:
```python
def set_PRUNED_to_value(study, value):
    trials = []
    for t in study.get_trials():
        if t.state == optuna.trial.TrialState.PRUNED:
            t = optuna.trial.create_trial(
                state=optuna.trial.TrialState.COMPLETE,
                value=value,
                params=t.params, 
                distributions=t.distributions,
                user_attrs=t.user_attrs, 
                system_attrs=t.system_attrs,
                intermediate_values=t.intermediate_values)
            trials.append(t)
        elif t.state == optuna.trial.TrialState.COMPLETE:
            trials.append(t)
        else:
            pass
    new_study=optuna.create_study()
    new_study.add_trials(trials)
    return new_study
```
as a method to the study object to transform a study object into an imputed study object.

**Comment by mreitschuster** [comment]
As there seems to be no interested, I am closing the issue. The piece of code seems good enough for those who need to include the pruned trials in the feature importnace analysis.

**Comment by raquelhortab** [comment]
@mreitschuster 
Hi, I got here searching for info about how pruned trials are taken into account. I have observed that the reported "best trial" in a study might not actually be the best. I believe the best intermediate value should be considered the best overall.  

I observed that in my studies, the one reported to be the best is not the actual best (see screenshot). It turns out the best one (brown line) was pruned. So I ask:

1. are intermediate values taken into account when optimizing the metrics? if a trial is pruned, will its results be taken into account? I really think they should, otherwise we either choose to discard potentially good trials or we choose to not use pruning at all, which is not efficient
2. shouldn't the reported best trial be the one with actual best results?

![image](https://github.com/optuna/optuna/assets/19463905/a1fa7d7d-9c11-4aec-9b95-8ea79f3d1c00)
